commands/command_filter_process.go

```diff
@@ -0,0 +1,128 @@
+package commands
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/github/git-lfs/git"
+	"github.com/github/git-lfs/lfs"
+	"github.com/spf13/cobra"
+)
+
+const (
+	// cleanFilterBufferCapacity is the desired capacity of the
+	// `*git.PacketWriter`'s internal buffer when the filter protocol
+	// dictates the "clean" command. 512 bytes is (in most cases) enough to
+	// hold an entire LFS pointer in memory.
+	cleanFilterBufferCapacity = 512
+
+	// smudgeFilterBufferCapacity is the desired capacity of the
+	// `*git.PacketWriter`'s internal buffer when the filter protocol
+	// dictates the "smudge" command.
+	smudgeFilterBufferCapacity = git.MaxPacketLength
+)
+
+// filterSmudgeSkip is a command-line flag owned by the `filter-process` command
+// dictating whether or not to skip the smudging process, leaving pointers as-is
+// in the working tree.
+var filterSmudgeSkip bool
+
+// filterSmudge is a gateway to the `smudge()` function and serves to bail out
+// immediately if the pointer decoded from "from" has no data (i.e., is empty).
+// This function, unlike the implementation found in the legacy smudge command,
+// only combines the `io.Reader`s when necessary, since the implementation
+// found in `*git.PacketReader` blocks while waiting for the following packet.
+func filterSmudge(from io.Reader, to io.Writer, filename string) error {
+	var pbuf bytes.Buffer
+	from = io.TeeReader(from, &pbuf)
+
+	ptr, err := lfs.DecodePointer(from)
+	if err != nil {
+		// If we tried to decode a pointer out of the data given to us,
+		// and the file was _empty_, write out an empty file in
+		// response. This occurs because when the clean filter
+		// encounters an empty file, and writes out an empty file,
+		// instead of a pointer.
+		//
+		// TODO(taylor): figure out if there is more data on the reader,
+		// and buffer that as well.
+		if len(pbuf.Bytes()) == 0 {
+			if _, cerr := io.Copy(to, &pbuf); cerr != nil {
+				Panic(cerr, "Error writing data to stdout:")
+			}
+			return nil
+		}
+
+		return err
+	}
+
+	lfs.LinkOrCopyFromReference(ptr.Oid, ptr.Size)
+
+	return smudge(to, ptr, filename, filterSmudgeSkip)
+}
+
+func filterCommand(cmd *cobra.Command, args []string) {
+	requireStdin("This command should be run by the Git filter process")
+	lfs.InstallHooks(false)
+
+	s := git.NewFilterProcessScanner(os.Stdin, os.Stdout)
+
+	if err := s.Init(); err != nil {
+		ExitWithError(err)
+	}
+	if err := s.NegotiateCapabilities(); err != nil {
+		ExitWithError(err)
+	}
+
+Scan:
+	for s.Scan() {
+		var err error
+		var w *git.PktlineWriter
+
+		req := s.Request()
+
+		s.WriteStatus(statusFromErr(nil))
```

Comment:
Hm. Maybe we can improve this part. This is how a typical request looks like:
```
packet:          git> command=smudge
packet:          git> pathname=path/testfile.dat
packet:          git> 0000
packet:          git> CONTENT
packet:          git> 0000
packet:          git< status=success         # (1) This success is L86 above.
packet:          git< 0000
packet:          git< SMUDGED_CONTENT
packet:          git< 0000
packet:          git< status=success         # (2) This status is set later.
packet:          git< 0000
```

Right now (1) will never set the status to `error`. Instead, GitLFS will just shutdown. Right now (2) sets the status to `error` in case of an error which makes Git shutdown GitLFS (because the filter has the `required=true` flag).

I think we should be consistent here and I think I prefer the (1) approach. If GitLFS detects an error then it should shutdown. This way we ensure that Git gets our "something is wrong" message very clearly.